src/appleseed/renderer/modeling/bsdf/hairbsdf.cpp

```diff
@@ -0,0 +1,840 @@
+            float cos_gamma_t = std::sqrt(std::max(0.0f, 1.0f - sin_gamma_t * sin_gamma_t));
+            float gamma_t = std::asin(clamp(sin_gamma_t, -1.0f, 1.0f));
+            float gamma_o = std::asin(clamp(values->m_precomputed.m_h, -1.0f, 1.0f));
+
+            // Compute attenuation for the single path through the hair cylinder
+            Spectrum T = exp(-values->m_precomputed.m_sigma_a * (( 2.f * cos_gamma_t) / cos_theta_t));
+
+            // Determine which scattering component to sample for hair scattering
+            std::array<Spectrum, 4> ap = attenuation(cos_theta_o, values->m_eta, values->m_precomputed.m_h, T);
+            std::array<float, 4> ap_pdf = attenuation_pdf(cos_theta_o, values->m_eta, values->m_precomputed.m_h, T);
+            int p;
+            for (p = 0; p < 3; ++p) {
+                if (sample_N[0] < ap_pdf[p]) break;
+                sample_N[0] -= ap_pdf[p];
+            }
+
+            sample_M[0] = std::max(sample_M[0], float(1e-5));
+            float theta_o = std::asin(clamp(sin_theta_o, -1.0f, 1.0f));
+
+            // Deviation of outgoing angle due to cuticular scales
+
+            float theta_o_R = theta_o - 2.0f * deg_to_rad(values->m_alpha);
+            float theta_o_TT = theta_o + deg_to_rad(values->m_alpha);
+            float theta_o_TRT = theta_o + 4.0f * deg_to_rad(values->m_alpha);
+
+            // Sample longitudinal function for given component of scattering
+
+            float sin_theta_i;
+            if (p == 0) {
+                sin_theta_i = sample_longitudinal(
+                    std::sin(theta_o_R),
+                    std::cos(theta_o_R),
+                    values->m_precomputed.m_v,
+                    sample_M);
+            }
+            if (p == 1) {
+                sin_theta_i = sample_longitudinal(
+                    std::sin(theta_o_TT),
+                    std::cos(theta_o_TT),
+                    values->m_precomputed.m_v * .25f,
+                    sample_M);
+            }
+            if (p == 2) {
+                sin_theta_i = sample_longitudinal(
+                    std::sin(theta_o_TRT),
+                    std::cos(theta_o_TRT),
+                    values->m_precomputed.m_v * 4.f,
+                    sample_M);
+            }
+            float cos_theta_i = std::sqrt(std::max(0.0f, 1.f - sin_theta_i * sin_theta_i));
+
+
+            // Sample azimuthal scattering function to compute change in azimuthal direction
+            float dphi;
+            dphi = calc_phi(p, gamma_o, gamma_t) +
+            sample_trimmed_logistic(sample_N[1], values->m_precomputed.m_s);
+            
+            // Compute wi from sampled hair scattering angles
+            float phi_i = phi_o + dphi;
+            Vector3f wi = Vector3f(sin_theta_i, cos_theta_i * std::sin(phi_i), cos_theta_i * std::cos(phi_i));
+
+            Spectrum bsdf_f(0.f);
+            float bsdf_pdf(0.f);
+
+            // RR contribution of scattering
+            bsdf_f +=
```

Comment:
A formatting suggestion to improve readability:

    bsdf_f +=
        ap[0] *
        longitudinal(
            cos_theta_i,
            std::abs(std::cos(theta_o_R)),
            sin_theta_i,
            std::sin(theta_o_R),
            values->m_precomputed.m_v) *
        azimuthal(dphi, 0, values->m_precomputed.m_s, gamma_o, gamma_t);